generator/templates/wire/WireClient.cpp

```diff
@@ -411,6 +510,49 @@ namespace wire {
                         return true;
                     }
                 {% endfor %}
+
+                bool HandleBufferMapReadAsyncCallback(const uint8_t** commands, size_t* size) {
+                    const auto* cmd = GetCommand<ReturnBufferMapReadAsyncCallbackCmd>(commands, size);
+                    if (cmd == nullptr) {
+                        return false;
+                    }
+
+                    auto* buffer = device->buffer.GetObject(cmd->bufferId);
+                    uint32_t bufferSerial = device->buffer.GetSerial(cmd->bufferId);
+
+                    //* The buffer might have been deleted or recreated so this isn't an error.
+                    if (buffer == nullptr || bufferSerial != cmd->bufferSerial) {
+                        return true;
+                    }
+
+                    //* The requests can have been deleted via an Unmap so this isn't an error.
+                    auto requestIt = buffer->readRequests.find(cmd->requestSerial);
+                    if (requestIt == buffer->readRequests.end()) {
+                        return true;
+                    }
+
+                    auto request = requestIt->second;
+
+                    //* On success, we copy the data locally because the IPC buffer isn't valid outside of this function
+                    if (cmd->status == NXT_BUFFER_MAP_READ_STATUS_SUCCESS) {
+
+                        //* The server didn't send the right amount of data, this is an error and could cause
+                        //* the application to crash if we did call the callback.
+                        if (request.size != cmd->dataLength) {
+                            return false;
+                        }
+
+                        if (buffer->mappedData != nullptr) {
+                            return false;
+                        }
+                        buffer->mappedData = malloc(request.size);
+                        memcpy(buffer->mappedData, cmd->GetData(), request.size);
+                    }
+
+                    request.callback(static_cast<nxtBufferMapReadStatus>(cmd->status), buffer->mappedData, request.userdata);
```

Comment:
Yes failed MapReadAsync always call with nullptr, because no value would make sense in an error case. The wire makes sure requests are not confused with one another with the "request serial" that uniquely identifies a request for a given buffer.

This is also tested (briefly) in BufferValidationTest.MapReadUnmapBeforeResultThenMapAgain